Upgrade to MACI 0.9.4 and x32 Circuits

daodesigner opened the pull request:
uses circuts pinned [here](https://gateway.pinata.cloud/ipfs/QmWSxPBNYDtsK23KwYdMtcDaJg3gWS3LBsqMnENrVG6nmc), updates tests, track bytecode size, nuanced contract optimization. 

closes #367
closes #366

daodesigner committed:
Upgrade to MACI 0.9.4 and x32 Circuits from ceremony, update tests, track bytecode size, nuanced contract optimization

xuhcc commented on contracts/contracts/snarkVerifiers/BatchUpdateStateTreeVerifier32.sol:
```diff
@@ -0,0 +1,254 @@
+            vk.delta2
+        );
+    }
+}
```
Please update `contracts/contracts/snarkVerifiers/README.md` with trusted setup details for x32 circuits

daodesigner replied:
will do

xuhcc commented on contracts/hardhat.config.ts:
```diff
@@ -39,15 +42,68 @@ const config: HardhatUserConfig = {
+          },
+        },
+      }
+    }
```
Why did you chose to optimize only these contracts?

daodesigner replied:
The solidity optimizer trades off between optimizing for bytecode size (deployment) by setting the number of runs to 0, or optimize for method gas costs (makes bytecode size and deployment more expensive) by setting the number of runs high. 

Both of the contracts are optimized: the ones with zero runs are optimized for deployment cost to get under the block gas limit, the ones with non zero runs are optimized for method gas costs. Both of these were required to have versions of the contracts that were under the 24kb max size and cost less than the block gas limit to deploy.

*NOTE: there is a max number of runs that can be verified on etherscan.

xuhcc replied:
>Both of these were required to have versions of the contracts that were under the 24kb max size and cost less than the block gas limit to deploy.

Yesterday you said that gas limit on xDai is 20M, so this should not be a problem.
Is `runs: 0` really needed to make `FundingRoundFactory` contract smaller than EIP170 size limit or the number of runs could be slightly increased to make `MACI` and `FundingRound` calls slightly cheaper?

daodesigner replied:
ok the following trades off around 1kb in size for factories for a few hundred runs.


```
solidity: {
    version: '0.6.12',
    settings: {
      optimizer: {
        enabled: true,
-        runs: 0,
+        runs: 200,
      },
    },
    overrides: {   
      "contracts/FundingRoundFactory.sol": {
        version: "0.6.12",
        settings: {
          optimizer: {
            enabled: true,
            runs: 1000000,
          },
        },
      },
      "contracts/FundingRound.sol": {
        version: "0.6.12",
        settings: {
          optimizer: {
            enabled: true,
            runs: 1000000,
          },
        },
      },
```
      
New contract sizes:

![image](https://user-images.githubusercontent.com/58278045/125662977-bba715b5-6792-482d-94d4-a02b2567cdb9.png)

xuhcc commented on contracts/hardhat.config.ts:
```diff
@@ -39,15 +42,68 @@ const config: HardhatUserConfig = {
       optimizer: {
         enabled: true,
-        runs: 20,
+        runs: 0,
```
Why you're setting default number of rounds to 0?

daodesigner replied:
The solidity optimizer trades off between optimizing for bytecode size (deployment) by setting the number of runs to 0 we optimize deployment costs.

xuhcc commented on contracts/utils/deployment.ts:
```diff
@@ -19,9 +19,9 @@ export function linkBytecode(
-    treeDepths: { stateTreeDepth: 4, messageTreeDepth: 4, voteOptionTreeDepth: 2 },
+    batchUstVerifier: 'BatchUpdateStateTreeVerifier32',
+    qvtVerifier: 'QuadVoteTallyVerifier32',
+    treeDepths: { stateTreeDepth: 32, messageTreeDepth: 32, voteOptionTreeDepth: 3 },
```
It's a test circuit config.
You should have added new config (e.g. 'big' or 'x32', depending on how it's called in MACI).

daodesigner replied:
will do

xuhcc commented:
Doesn't MACI 0.8 introduce new dependencies? If it does, documentation should be updated as well.

daodesigner commented:
Will make sure to update the docs as well.

daodesigner committed:
update readme, optimize solc optimizer runs, nits

daodesigner committed:
Merge pull request #377 from daodesigner/feature/scale-circuits-billion-users

update readme, optimize solc optimizer runs, nits

xuhcc commented on docs/trusted-setup.md:
````diff
@@ -60,3 +70,22 @@ Verify:
+
+Quadratic Vote Tally Verifier: 
+```
+sudo g++ -pthread main.cpp calcwit.cpp utils.cpp fr.cpp fr.o qvt32.c -o qvt32 -lgmp -std=c++11 -O3 -fopenmp -DSANITY_CHECK
````
sudo is not needed here

daodesigner replied:
will remove

xuhcc commented on docs/trusted-setup.md:
````diff
@@ -60,3 +70,22 @@ Verify:
+```
+BatchUst Verifier:
+```
+sudo g++ -pthread main.cpp calcwit.cpp utils.cpp fr.cpp fr.o batchUst.c -o batchUst -lgmp -std=c++11 -O3 -fopenmp -DSANITY_CHECK
````
```
cc1plus: fatal error: batchUst.c: No such file or directory
```

However, there's a file named `batchUst32.c` in trusted setup directory `QmWSxPBNYDtsK23KwYdMtcDaJg3gWS3LBsqMnENrVG6nmc`

daodesigner replied:
2 points to raise here: 

- this was useful incase the ceremony fails mid run, you can use the files in the trusted set up directory
- the trusted set up parameters are currently not permanent (they were generated by one person). The multiparty trusted ceremony will happen in three weeks at which point we'll have to update the contracts again.

remove or clarify?

xuhcc replied:
Remove

xuhcc commented on docs/trusted-setup.md:
````diff
@@ -60,3 +70,22 @@ Verify:
+Use v0.9.4 to install Zkutil use the script at maci/circuits/scripts/installZkutil.sh
+
+```
+./circuits/scripts/installZkutil.sh
````
Normally the path to `zkutil` binary is set with `zkutil_bin` config parameter. Installation is already described in `contracts/README.md`

daodesigner replied:
will remove

xuhcc reviewed (changes requested):
E2E tests fail with x32 circuits:

```
Progress: 1 / 2; batch index: 4
batchUst32: main.cpp:165: void iterateArr(Circom_CalcWit*, int, Circom_Sizes, json, ItFunc): Assertion `(sizes[0] == 1)&&(sizes[1] == 0)' failed.
Loading circuit from clrfund/contracts/snark-params/batchUst32.r1cs...
thread 'main' panicked at 'unable to open.: Os { code: 2, kind: NotFound, message: "No such file or directory" }', zkutil-0.3.2/src/circom_circuit.rs:493:10
```

xuhcc commented on docs/coordinator.md:
```diff
@@ -2,22 +2,30 @@
 Follow instructions in README.md to install necessary dependencies.
 
+### Medium Circuits
 Download [zkSNARK parameters](https://gateway.pinata.cloud/ipfs/QmRzp3vkFPNHPpXiu7iKpPqVnZB97wq7gyih2mp6pa5bmD) for 'medium' circuits into `circuits/params/` directory and rebuild the keys:
```
Are these parameters compatible with MACI 0.9.4?

If not, this information should be removed

daodesigner replied:
will remove

daodesigner commented:
> E2E tests fail with x32 circuits:
> 
> ```
> Progress: 1 / 2; batch index: 4
> batchUst32: main.cpp:165: void iterateArr(Circom_CalcWit*, int, Circom_Sizes, json, ItFunc): Assertion `(sizes[0] == 1)&&(sizes[1] == 0)' failed.
> Loading circuit from clrfund/contracts/snark-params/batchUst32.r1cs...
> thread 'main' panicked at 'unable to open.: Os { code: 2, kind: NotFound, message: "No such file or directory" }', zkutil-0.3.2/src/circom_circuit.rs:493:10
> ```
will try to recreate on the server with maci 9.4 and your environment, I've also had trouble running the proof generation locally, and have been debugging with @weijiekoh, could you let me know what version you have installed for: Ubuntu, libc6, zkutil? (usual suspects) so I can recreate with your environment?

Also, is this the only failing test on your end? Last I ran the test harness everything looked green: 
![image](https://user-images.githubusercontent.com/58278045/125724263-d8610337-4cfc-4266-b61a-2a73298d5d2e.png)

xuhcc commented on docs/trusted-setup.md:
````diff
@@ -4,23 +4,33 @@
+```
+cd circuits
+./scripts/buildSnarks32.sh
+```
````
Files `batchUst32` and `qvt32` must be flagged as executable after downloading, at least in "using clrfund scripts" scenario.

```
chmod a+x snark-params/batchUst32
```

(By the way, we don't use MACI CLI directly anymore, only for debugging)

xuhcc replied:
Not addressed.

xuhcc commented:
>could you let me know what version you have installed for: Ubuntu, libc6, zkutil? (usual suspects) so I can recreate with your environment?

Just try to run E2E tests with x32 circuits.

The error comes from zkutil, I don't think it is related to any of the system libraries. Looks like upstream bug (cc @weijiekoh)

E2E tests are also failing with 'test' circuits because you didn't update parameters for them, but that's another issue.

>Also, is this the only failing test on your end? Last I ran the test harness everything looked green:

All units tests are passing.

weijiekoh commented:
My bad, I compiled the witness generation binaries using a new version of Ubuntu which updated glibc.

I've recompiled these binaries with an older version. Please try these: https://www.dropbox.com/sh/u6qaqlgzwh2un8z/AABjtfK7WWzdMGwn8TuGyAq-a?dl=0

The other files (.sol, .bin, .dat, etc) are not affected.

xuhcc commented:
@weijiekoh Tried these binaries, but still getting the same error:

```
Progress: 1 / 2; batch index: 4
batchUst32: maci/circuits/params/main.cpp:165: void iterateArr(Circom_CalcWit*, int, Circom_Sizes, json, ItFunc): Assertion `(sizes[0] == 1)&&(sizes[1] == 0)' failed.
Loading circuit from clrfund/contracts/snark-params/batchUst32.r1cs...
thread 'main' panicked at 'unable to open.: Os { code: 2, kind: NotFound, message: "No such file or directory" }', zkutil-0.3.2/src/circom_circuit.rs:493:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: unable to compute batch update state tree witness data
```

I also tried to compile them myself, with the same result.

I have glibc 2.33 in my system.

weijiekoh commented:
@daodesigner what were the parameters set for max users, max messages, etc?

xuhcc commented:
I've changed batch size from 4 to 8 and assertion error disappeared.
However, zkutil still fails when attempting to read witness file https://github.com/poma/zkutil/blob/v0.3.2/src/circom_circuit.rs#L493

daodesigner commented:
> I've changed batch size from 4 to 8 and assertion error disappeared.
> 
> However, zkutil still fails when attempting to read witness file https://github.com/poma/zkutil/blob/v0.3.2/src/circom_circuit.rs#L493

Could you verify the other parameters, and versions of things you have installed as well? 8 should be correct for batch size. Will try to recreate your error today, have a few thoughts on what it might be.

Also could you clarify on the change needed for test circuits? As you asked I made a new param config for x32 circuits rather than replace the one you had for test circuits per your other cr above. 

In any case, have a new server running with the details you've given me so far and will ping back here this afternoon once I know more.

daodesigner commented on contracts/utils/deployment.ts:
```diff
@@ -33,6 +33,11 @@ const CIRCUITS: {[name: string]: any} = {
+  x32: {
+    batchUstVerifier: 'BatchUpdateStateTreeVerifier32',
+    qvtVerifier: 'QuadVoteTallyVerifier32',
+    treeDepths: { stateTreeDepth: 32, messageTreeDepth: 32, voteOptionTreeDepth: 3 },
```
X32 default circuit params

xuhcc commented:
@daodesigner

>Could you verify the other parameters, and versions of things you have installed as well?

Verify what exactly? I've downloaded x32 snark parameters according to the instructions you provided in this PR and run the tests.

>Also could you clarify on the change needed for test circuits?

I'm talking about verifiers for 'test' and 'medium' circuits in `contracts/contracts/snarkVerifiers` and corresponding snark parameters which were generated for MACI 0.7. We'll decide later what to do with them (re-generate or remove).

daodesigner reviewed (commented):
> @daodesigner
> 
> > Could you verify the other parameters, and versions of things you have installed as well?
> 
> Verify what exactly? I've downloaded x32 snark parameters according to the instructions you provided in this PR and run the tests.
> 
> > Also could you clarify on the change needed for test circuits?
> 
> I'm talking about verifiers for 'test' and 'medium' circuits in `contracts/contracts/snarkVerifiers` and corresponding snark parameters which were generated for MACI 0.7. We'll decide later what to do with them (re-generate or remove).

Ok, so I wasn't able to recreate your issue on my end: 


![image](https://user-images.githubusercontent.com/58278045/125836706-1163df2f-9c29-474e-9418-a31f4c216410.png)
maybe check the path you are exporting for snark params and make it absolute just to be sure?  `export NODE_CONFIG='{"snarkParamsPath": "/media/contracts/snark-params/", "zkutil_bin": "/usr/bin/zkutil"}'`


but I am getting a revert on the batch process message after the proofs are generated (correctly), will dig into the E13  error now.


![image](https://user-images.githubusercontent.com/58278045/125836935-ec00f3f5-6014-42f5-b3ce-7730989c968b.png)

UPDATE:
The E13 revert has to do with the voting period (not related to MACI implementation details) 
```   
string constant ERROR_VOTING_PERIOD_PASSED = "E13";
string constant ERROR_VOTING_PERIOD_NOT_OVER = "E13";
```

will try to figure out which it is (I think it's probably the second),  fix the test rig , and add it so it's run with the other tests

daodesigner commented:
> will try to figure out which it is (I think it's probably the second), fix the test rig , and add it so it's run with the other tests

was almost certainly the second, just as a sanity check looks like things are fine on the MACI side: 
`yarn tally:local`
![image](https://user-images.githubusercontent.com/58278045/125847889-7bbe1045-7905-44c6-a40a-21a3249149d7.png)
`yarn finalize:local`
![image](https://user-images.githubusercontent.com/58278045/125847991-02ebae56-2b83-4277-a27e-85b1dafc990d.png)
`yarn claim:local`
![image](https://user-images.githubusercontent.com/58278045/125848058-f561be85-6282-42ad-b196-5bf7415251d0.png)

I will take a closer look at the e2e test rig this afternoon and then follow up to make sure we understand where the error is coming from.

Should also mention the gas costs for some of these transactions did decrease so good call on tweaks for the optimizer @xuhcc.

daodesigner commented:
> I will take a closer look at the e2e test rig this afternoon and then follow up to make sure we understand where the error is coming from.

E2E tests passing, seems I introduced a small bug in the test rig when I addressed the previous code review.

![image](https://user-images.githubusercontent.com/58278045/125861670-bb6dc9b3-ec4d-4d22-8244-f6edc7b2da60.png)

daodesigner committed:
nit docs, bump up gas for optimizer deploy tradeoff, fix e2e tests

xuhcc commented:
Latest commit doesn't fix E2E tests (`No such file or directory`)

daodesigner commented:
> Latest commit doesn't fix E2E tests (`No such file or directory`)

Could you send me the full trace for the error you’re getting and versions for Ubuntu/package deps you have installed `libc libgmp-dev nlohmann-json3-dev nasm g++ nodejs npm cargo rust zkutil` and the `NODE_CONFIG` you exported? Will need a this information to recreate what you are seeing.

Not sure why it fails in your environment but it runs fine on the server.

@weijiekoh any chance we could give @xuhcc server credentials to run the tests for the code review there?

xuhcc commented:
@weijiekoh Does MACI CLI check the return code of `batchUst32`? For example, why it continued to execute after assertion error ```batchUst32: main.cpp:165: void iterateArr(Circom_CalcWit*, int, Circom_Sizes, json, ItFunc): Assertion `(sizes[0] == 1)&&(sizes[1] == 0)' failed.```?
If it does not check return code, I would start with fixing that. It is likely that the actual error is hidden because of that.

@daodesigner Are `libc libgmp-dev nlohmann-json3-dev nasm` needed for generating proofs? This is the first time you're mentioning them. Actually these packages look like a compile-time dependencies, but I'm using precompiled binaries.

I don't use Ubuntu (I use Arch). Ubuntu must not be a requirement.

zkutil 0.3.2, node v12.22.0, `NODE_CONFIG='{"snarkParamsPath": "../../../contracts/snark-params/", "zkutil_bin": "/usr/bin/zkutil"}`

Here's a stacktrace:

```
   at Object.error (clrfund/node_modules/shelljs/src/common.js:110:27)
    at execSync (clrfund/node_modules/shelljs/src/exec.js:102:12)
    at Object._exec (clrfund/node_modules/shelljs/src/exec.js:205:12)
    at Object.exec (clrfund/node_modules/shelljs/src/common.js:335:23)
    at clrfund/node_modules/maci-circuits/ts/index.ts:242:11
    at step (clrfund/node_modules/maci-circuits/build/index.js:33:23)
    at Object.next (clrfund/node_modules/maci-circuits/build/index.js:14:53)
    at clrfund/node_modules/maci-circuits/build/index.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (clrfund/node_modules/maci-circuits/build/index.js:4:12)
    at genProofAndPublicSignals (clrfund/node_modules/maci-circuits/build/index.js:211:12)
    at Object.genBatchUstProofAndPublicSignals (clrfund/node_modules/maci-circuits/ts/index.ts:131:12)
    at clrfund/node_modules/maci-cli/ts/genProofs.ts:265:28
    at step (clrfund/node_modules/maci-cli/build/genProofs.js:33:23)
    at Object.next (clrfund/node_modules/maci-cli/build/genProofs.js:14:53)
    at fulfilled (clrfund/node_modules/maci-cli/build/genProofs.js:5:58)
```

xuhcc commented:
>Does MACI CLI check the return code of batchUst32? For example, why it continued to execute after assertion error batchUst32: main.cpp:165: void iterateArr(Circom_CalcWit*, int, Circom_Sizes, json, ItFunc): Assertion `(sizes[0] == 1)&&(sizes[1] == 0)' failed.? If it does not check return code, I would start with fixing that. It is likely that the actual error is hidden because of that.

I figured it out. The return code is 0. The binary pretends that everything is fine but doesn't actually generate witness JSON.

```
batchUst32 1626417004245.input.json 1626417004245.witness.json
```

daodesigner commented:
> @daodesigner Are `libc libgmp-dev nlohmann-json3-dev nasm` needed for generating proofs? This is the first time you're mentioning them. Actually these packages look like a compile-time dependencies, but I'm using precompiled binaries.
> I don't use Ubuntu (I use Arch). Ubuntu must not be a requirement.

Correct, the binaries only require the correct version of glibc to excute. Older Ubuntu distros don't come packaged with glibc 2.3.3 which you need to run the executables pinned on IPFS. If you have an older libc version you need to recompile the binaries- in this case you need the dependencies I mentioned to be up to date (WJ also linked binaries compiled with an older glibc version via drop box above). This shouldn't be the issue if you're using Arch, but good to cross it out.

Thank you for the details you provided.

> I figured it out. The return code is 0. The binary pretends that everything is fine but doesn't actually generate witness JSON.
>
> `batchUst32 1626417004245.input.json 1626417004245.witness.json`

I couldn't replicate this either, witness JSON is generated fine on my end, are the input.json files also missing in your runs?
![image](https://user-images.githubusercontent.com/58278045/125908205-4dcf7c49-6cf4-4138-b4c6-1fc06bfa3ab6.png)

xuhcc commented:
>The return code is 0.

Hmm that's what `shell.exec` is saying. But if I execute binary with the same paramters via real shell it prints error and returns code 132:

```
./batchUst32 1626417004245.input.json 1626417004245.witness.json
Up to loadJson 0.02428100000000000050
Items : 21
Illegal instruction (core dumped)
```

xuhcc commented:
@weijiekoh 

Here's the gdb output, hopefully it will help:

```
(gdb) bt full
#0  0x00005555555799b2 in Fr_rawMMul1 ()
No symbol table info available.
#1  0x000000000000001a in ?? ()
No symbol table info available.
#2  0x000000000000001b in ?? ()
No symbol table info available.
#3  0x00000000000001f5 in ?? ()
No symbol table info available.
#4  0x00000000000001f6 in ?? ()
No symbol table info available.
#5  0x0000555555579e3a in posMontgomeryShort ()
No symbol table info available.
#6  0x000055555557a0af in add_s1nl2m ()
No symbol table info available.
#7  0x00007fffffffb960 in ?? ()
No symbol table info available.
#8  0x00007ffff50424f8 in ?? ()
No symbol table info available.
#9  0x0000000000000000 in ?? ()
No symbol table info available.
```

daodesigner commented:
@xuhcc any progress here? I don't know that there is anything further I can do on my end to push this along; tests are all passing on Ubuntu servers we are using and I've been unable to replicate what you are seeing. Kindly, could you either decline the PR or accept it but create an issue for generating proofs using your system specs? just want some clarity on the timeframe here to plan accordingly for the next sprint.

xuhcc commented:
I haven't got a response from @weijiekoh yet

>I don't know that there is anything further I can do on my end to push this along; tests are all passing on Ubuntu servers we are using and I've been unable to replicate what you are seeing.

It's not acceptable that it only works on your server. Ideally, clrfund should work on all major OSes. Everyone should be able to run and coordinate rounds. At the very least, it must work on Linux. Also, E2E tests are essential for development.

If you want to get this merged, you need to resolve this issue somehow (e.g. by fixing the bug or by providing instructions on how to run it outside your server).

daodesigner commented:
> @weijiekoh 
> 
> 
> 
> Here's the gdb output, hopefully it will help:
> 
> 
> 
> ```
> 
> (gdb) bt full
> 
> #0  0x00005555555799b2 in Fr_rawMMul1 ()
> 
> No symbol table info available.
> 
> #1  0x000000000000001a in ?? ()
> 
> No symbol table info available.
> 
> #2  0x000000000000001b in ?? ()
> 
> No symbol table info available.
> 
> #3  0x00000000000001f5 in ?? ()
> 
> No symbol table info available.
> 
> #4  0x00000000000001f6 in ?? ()
> 
> No symbol table info available.
> 
> #5  0x0000555555579e3a in posMontgomeryShort ()
> 
> No symbol table info available.
> 
> #6  0x000055555557a0af in add_s1nl2m ()
> 
> No symbol table info available.
> 
> #7  0x00007fffffffb960 in ?? ()
> 
> No symbol table info available.
> 
> #8  0x00007ffff50424f8 in ?? ()
> 
> No symbol table info available.
> 
> #9  0x0000000000000000 in ?? ()
> 
> No symbol table info available.
> 
> ```

please try to recompile the binaries on your machine with the -ggdb flag then get the logs output again. These logs are borked unless @weijiekoh can debug binaries directly. Again, the end to end tests are passing on the server but happy to help debug what ever is going on here.

auryn-macmillan commented:
Merging this PR to keep things moving.
